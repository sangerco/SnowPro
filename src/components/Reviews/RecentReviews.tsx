import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import {
  Dimmer,
  Loader,
  Rating,
  Header,
  Divider,
  Grid,
  Card,
  Segment,
} from "semantic-ui-react";
import { RootState, AppDispatch } from "../../redux/store";
import { fetchAllReviews } from "../../redux/slices/reviewSlice";
import { fetchSkiAreas } from "../../redux/slices/skiAreaSlice";

const RecentReviews: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const reviews = useSelector((state: RootState) => state.reviews);
  useEffect(() => {
    dispatch(fetchAllReviews());
  }, [dispatch]);

  useEffect(() => {
    dispatch(fetchSkiAreas());
  }, [dispatch]);

  const skiAreaState = useSelector((state: RootState) => state.skiAreas);
  const skiAreas = skiAreaState.skiAreas;

  const formatDate = (date: Date) => {
    const newDate = new Date(date);
    const options: Intl.DateTimeFormatOptions = {
      year: "numeric",
      month: "long",
      day: "numeric",
    };
    return newDate.toLocaleDateString("en-US", options);
  };

  if (reviews.error) {
    return (
      <Dimmer active>
        <div>
          <p>Error! Review cannot be retrieved! {`${reviews.error}`}</p>
        </div>
      </Dimmer>
    );
  }

  if (reviews.loading) {
    return (
      <Dimmer active>
        <Loader>Loading...</Loader>
      </Dimmer>
    );
  }

  if (Array.isArray(reviews)) {
    return <div>Error! Reviews cannot be retrieved!</div>;
  } else if (reviews) {
    return (
      <Grid>
        <Grid.Row>
          <Grid.Column width={4}>
            {skiAreaState.loading ? (
              <Dimmer active>
                <Loader>Loading...</Loader>
              </Dimmer>
            ) : skiAreaState.error ? (
              <Dimmer active>
                <Header as="h1">
                  Error! Ski Area Data cannot be retrieved! {skiAreaState.error}
                </Header>
              </Dimmer>
            ) : skiAreas && skiAreas.length > 0 ? (
              <Segment style={{ overflow: "auto", maxHeight: "75vh" }}>
                <Card id="ski-area-card">
                  {" "}
                  {/* max height overflow auto */}
                  {skiAreas.map((sa) => (
                    <Card.Content key={sa.slug}>
                      <Link to={`/ski-areas/${sa.slug}`}>{sa.name}</Link>
                    </Card.Content>
                  ))}
                </Card>
              </Segment>
            ) : null}
          </Grid.Column>
          <Grid.Column width={8}>
            <Card.Group>
              {reviews.reviews && reviews.reviews.length > 0 ? (
                <>
                  {reviews.reviews.map((review) => (
                    <Card>
                      <Card.Content key={review.id} id="review-card">
                        <Card.Header>
                          <Link to={`/ski-areas/reviews/${review.id}`}>
                            {review.header}
                          </Link>{" "}
                          <Divider />
                          {review.skiAreaName}
                        </Card.Header>
                        <Card.Description>
                          By {review.username}
                        </Card.Description>
                        <Card.Meta>
                          <Rating
                            icon="star"
                            defaultRating={review.stars}
                            maxRating={5}
                            disabled
                          />
                        </Card.Meta>
                        {formatDate(review.createdAt)}
                      </Card.Content>
                    </Card>
                  ))}
                </>
              ) : null}
            </Card.Group>
          </Grid.Column>
          <Grid.Column width={4}></Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }

  return null;
};

export default RecentReviews;
