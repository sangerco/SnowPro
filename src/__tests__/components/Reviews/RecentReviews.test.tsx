/* eslint-disable testing-library/no-node-access */
import React from "react";
import RecentReviews from "../../../components/Reviews/RecentReviews";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import nock from "nock";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test Recent Reviews", () => {
  const store = mockStore({
    reviews: {
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review 2",
          body: "test review body 2",
          stars: 1,
          photos: [],
          createdAt: "fake created date",
          replies: [],
        },
      ],
    },
    skiAreas: {
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  });
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <RecentReviews />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <RecentReviews />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("should show reviews", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({ "access-control-allow-origin": "*" })
      .get("/ski-areas")
      .reply(200, [
        {
          slug: "vail",
          name: "Vail",
        },
        {
          slug: "arapahoe-basin",
          name: "Arapahoe Basin",
        },
      ]);
    nock("http://localhost:5000")
      .defaultReplyHeaders({ "access-control-allow-origin": "*" })
      .get("/ski-areas/reviews")
      .reply(200, [
        {
          id: "review-1",
          header: "test review",
          skiAreaName: "test ski area",
          username: "testuser",
          stars: 5,
          createdAt: "12/11/22 12:22:22.00",
        },
        {
          id: "review-2",
          header: "test review 2",
          skiAreaName: "test ski area 2",
          username: "testuser2",
          stars: 1,
          createdAt: "12/12/22 12:22:22.00",
        },
      ]);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <RecentReviews />
        </BrowserRouter>
      </Provider>
    );

    expect(nock.pendingMocks().length).toBe(2);

    await waitFor(() => {
      const review = screen.getAllByTestId("review");
      expect(review[0]).toBeInTheDocument();
    });

    await waitFor(() => {
      const review = screen.getAllByTestId("review");
      expect(review[1]).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(nock.isDone()).toBe(true);
    });
  });
});
