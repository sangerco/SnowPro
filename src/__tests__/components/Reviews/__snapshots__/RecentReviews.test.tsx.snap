// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Recent Reviews should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui grid"
  >
    <div
      class="row"
    >
      <div
        class="four wide column"
      >
        <div
          class="ui segment"
          style="overflow: auto; max-height: 75vh;"
        >
          <div
            class="ui card"
            id="ski-area-card"
          >
             
            <div
              class="content"
            >
              <a
                href="/ski-areas/ski-area-1"
              >
                Ski Area 1
              </a>
            </div>
            <div
              class="content"
            >
              <a
                href="/ski-areas/ski-area-1"
              >
                Ski Area 1
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="eight wide column"
      >
        <div
          class="ui cards"
        >
          <div
            class="ui card"
          >
            <div
              class="content"
              data-testid="review"
              id="review-card"
            >
              <div
                class="header"
              >
                <a
                  href="/ski-areas/reviews/review-1"
                >
                  test review
                </a>
                 
                <div
                  class="ui divider"
                />
                Ski Area 1
              </div>
              <div
                class="description"
              >
                By testuser
              </div>
              <div
                class="meta"
              >
                <div
                  class="ui star disabled rating"
                  role="radiogroup"
                  tabindex="0"
                >
                  <i
                    aria-checked="false"
                    aria-posinset="1"
                    aria-setsize="5"
                    class="active icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="false"
                    aria-posinset="2"
                    aria-setsize="5"
                    class="active icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="false"
                    aria-posinset="3"
                    aria-setsize="5"
                    class="active icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="false"
                    aria-posinset="4"
                    aria-setsize="5"
                    class="active icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="true"
                    aria-posinset="5"
                    aria-setsize="5"
                    class="active icon"
                    role="radio"
                    tabindex="-1"
                  />
                </div>
              </div>
              Invalid Date
            </div>
          </div>
          <div
            class="ui card"
          >
            <div
              class="content"
              data-testid="review"
              id="review-card"
            >
              <div
                class="header"
              >
                <a
                  href="/ski-areas/reviews/review-2"
                >
                  test review 2
                </a>
                 
                <div
                  class="ui divider"
                />
                Ski Area 2
              </div>
              <div
                class="description"
              >
                By testuser2
              </div>
              <div
                class="meta"
              >
                <div
                  class="ui star disabled rating"
                  role="radiogroup"
                  tabindex="0"
                >
                  <i
                    aria-checked="true"
                    aria-posinset="1"
                    aria-setsize="5"
                    class="active icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="false"
                    aria-posinset="2"
                    aria-setsize="5"
                    class="icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="false"
                    aria-posinset="3"
                    aria-setsize="5"
                    class="icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="false"
                    aria-posinset="4"
                    aria-setsize="5"
                    class="icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="false"
                    aria-posinset="5"
                    aria-setsize="5"
                    class="icon"
                    role="radio"
                    tabindex="-1"
                  />
                </div>
              </div>
              Invalid Date
            </div>
          </div>
        </div>
      </div>
      <div
        class="four wide column"
      />
    </div>
  </div>
</DocumentFragment>
`;
